File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/ui/widgets/BattleStatusOverviewWidget.java
package de.limbusdev.guardianmonsters.battle.ui.widgets;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.actions.Actions;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.Array;

import de.limbusdev.guardianmonsters.guardians.battle.BattleSystem;
import de.limbusdev.guardianmonsters.guardians.battle.CombatTeam;
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian;
import de.limbusdev.guardianmonsters.ui.Constant;
import de.limbusdev.utils.geometry.IntVec2;


/**
 * Widget for displaying monster status in battle: HP, MP, EXP, Name, Level
 * HINT: Don't forget calling the init() method
 * Created by Georg Eckert 2016
 */
public class BattleStatusOverviewWidget extends BattleWidget
{

    private Array<MonsterStateWidget>
        monsterStateWidgetsLeft,
        monsterStateWidgetsRight;

    public BattleStatusOverviewWidget(Skin skin)
    {
        super();
        this.monsterStateWidgetsLeft = new Array<>();
        this.monsterStateWidgetsRight = new Array<>();

        // Hero Team ###############################################################################
        MonsterStateWidget msw = new MonsterStateWidget(skin, true);
        msw.setPosition(IndPos.statWPos1left.x, IndPos.statWPos1left.y, Align.topLeft);
        monsterStateWidgetsLeft.add(msw);
        msw = new MonsterStateWidget(skin, true);
        msw.setPosition(IndPos.statWPos2left.x, IndPos.statWPos2left.y, Align.topLeft);
        monsterStateWidgetsLeft.add(msw);
        msw = new MonsterStateWidget(skin, true);
        msw.setPosition(IndPos.statWPos3left.x, IndPos.statWPos3left.y, Align.topLeft);
        monsterStateWidgetsLeft.add(msw);

        // Opponent Team ###########################################################################
        msw = new MonsterStateWidget(skin, false);
        msw.setPosition(IndPos.statWPos1right.x, IndPos.statWPos1right.y, Align.topRight);
        monsterStateWidgetsRight.add(msw);
        msw = new MonsterStateWidget(skin, false);
        msw.setPosition(IndPos.statWPos2right.x, IndPos.statWPos2right.y, Align.topRight);
        monsterStateWidgetsRight.add(msw);
        msw = new MonsterStateWidget(skin, false);
        msw.setPosition(IndPos.statWPos3right.x, IndPos.statWPos3right.y, Align.topRight);
        monsterStateWidgetsRight.add(msw);

        for(MonsterStateWidget w : monsterStateWidgetsLeft) {addActor(w);}
        for(MonsterStateWidget w : monsterStateWidgetsRight) {addActor(w);}

        setDebug(Constant.DEBUGGING_ON, true);

    }

    private void addStatusWidgetsForTeam(CombatTeam team, boolean side)
    {
        Array<MonsterStateWidget> stateWidgets = side ?
            monsterStateWidgetsLeft : monsterStateWidgetsRight;

        // Clear Actions
        for(MonsterStateWidget w : stateWidgets)
        {
            w.clearActions();
            w.remove();
            w.setVisible(true);
            w.setColor(Color.WHITE);
        }

        // Initialize UI
        for(int key : team.keys())
        {
            stateWidgets.get(key).init(team.get(key));
            addActor(stateWidgets.get(key));
        }
    }

    public void init(BattleSystem battleSystem)
    {
        addStatusWidgetsForTeam(battleSystem.getQueue().getCombatTeamLeft(), Constant.LEFT);
        addStatusWidgetsForTeam(battleSystem.getQueue().getCombatTeamRight(), Constant.RIGHT);
    }

    public void updateStatusWidgetToSubstitute(int pos, boolean side, AGuardian guardian)
    {
        Array<MonsterStateWidget> stateWidgets = side ?
            monsterStateWidgetsLeft : monsterStateWidgetsRight;

        MonsterStateWidget w = stateWidgets.get(pos);
        w.clearActions();
        w.addAction(Actions.sequence(
            Actions.fadeOut(1f),
            Actions.run(() -> {w.init(guardian);}),
            Actions.fadeIn(1f)
        ));
    }


    /**
     * Possible Indicator coordinates
     */
    private final static class IndPos
    {
        private static final IntVec2 statWPos1left = new IntVec2(56+24, 360-24);
        private static final IntVec2 statWPos2left = new IntVec2(56, 360-48);
        private static final IntVec2 statWPos3left = new IntVec2(56+48, 360);
        private static final IntVec2 statWPos1right = new IntVec2(640-8-24, statWPos1left.y);
        private static final IntVec2 statWPos2right = new IntVec2(640-8, statWPos2left.y);
        private static final IntVec2 statWPos3right = new IntVec2(640-8-48, statWPos3left.y);
    }


}

File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/ui/widgets/BattleStatusOverviewWidget.kt
package de.limbusdev.guardianmonsters.battle.ui.widgets

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.actions.Actions
import com.badlogic.gdx.utils.Align
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ArrayMap
import de.limbusdev.guardianmonsters.battle.BattleDebugger

import de.limbusdev.guardianmonsters.guardians.Side
import de.limbusdev.guardianmonsters.guardians.battle.BattleSystem
import de.limbusdev.guardianmonsters.guardians.battle.CombatTeam
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian
import de.limbusdev.utils.extensions.set
import de.limbusdev.utils.geometry.IntVec2
import ktx.actors.then


/**
 * Widget for displaying monster status in battle: HP, MP, EXP, Name, Level
 * HINT: Don't forget calling the initialize() method
 * Created by Georg Eckert 2016
 */
class BattleStatusOverviewWidget() : BattleWidget()
{
    // .................................................................................. Properties
    private val monsterStateWidgets : ArrayMap<Side, Array<MonsterStateWidget>> = ArrayMap()


    // ................................................................................ Constructors
    init
    {
        monsterStateWidgets[Side.LEFT]  = Array()
        monsterStateWidgets[Side.RIGHT] = Array()

        // Hero Team
        monsterStateWidgets[Side.LEFT].add(setupGuardianStateWidget(true, IndPos.statWPos1left, Side.LEFT))
        monsterStateWidgets[Side.LEFT].add(setupGuardianStateWidget(true, IndPos.statWPos2left, Side.LEFT))
        monsterStateWidgets[Side.LEFT].add(setupGuardianStateWidget(true, IndPos.statWPos3left, Side.LEFT))

        // Opponent Team
        monsterStateWidgets[Side.RIGHT].add(setupGuardianStateWidget(true, IndPos.statWPos1right, Side.RIGHT))
        monsterStateWidgets[Side.RIGHT].add(setupGuardianStateWidget(true, IndPos.statWPos2right, Side.RIGHT))
        monsterStateWidgets[Side.RIGHT].add(setupGuardianStateWidget(true, IndPos.statWPos3right, Side.RIGHT))

        for (w in monsterStateWidgets[Side.LEFT])  { addActor(w) }
        for (w in monsterStateWidgets[Side.RIGHT]) { addActor(w) }

        setDebug(BattleDebugger.SCENE2D_DEBUG, true)
    }

    private fun setupGuardianStateWidget(showEXP: Boolean, position: IntVec2, side: Side) : MonsterStateWidget
    {
        val align = when(side) { Side.LEFT -> Align.topLeft; Side.RIGHT -> Align.topRight }
        val widget = MonsterStateWidget(showEXP)
        widget.setPosition(position.xf, position.yf, align)
        return widget
    }

    private fun addStatusWidgetsForTeam(team: CombatTeam, side: Side)
    {
        // Clear Actions
        for (w in monsterStateWidgets[side]) {
            w.clearActions()
            w.remove()
            w.isVisible = true
            w.color = Color.WHITE
        }

        // Initialize UI
        for (key in team.keys()) {
            monsterStateWidgets[side].get(key).initialize(team.get(key))
            addActor(monsterStateWidgets[side].get(key))
        }
    }

    fun initialize(battleSystem: BattleSystem)
    {
        addStatusWidgetsForTeam(battleSystem.queue.combatTeamLeft, Side.LEFT)
        addStatusWidgetsForTeam(battleSystem.queue.combatTeamRight, Side.RIGHT)
    }

    fun updateStatusWidgetToSubstitute(pos: Int, side: Side, guardian: AGuardian)
    {
        val w = monsterStateWidgets[side].get(pos)
        w.clearActions()
        w.addAction(

                Actions.fadeOut(1f)                     then
                Actions.run { w.initialize(guardian) }  then
                Actions.fadeIn(1f)
        )
    }


    /**
     * Possible Indicator coordinates
     */
    private object IndPos
    {
        val statWPos1left = IntVec2(56 + 24, 360 - 24)
        val statWPos2left = IntVec2(56, 360 - 48)
        val statWPos3left = IntVec2(56 + 48, 360)
        val statWPos1right = IntVec2(640 - 8 - 24, statWPos1left.y)
        val statWPos2right = IntVec2(640 - 8, statWPos2left.y)
        val statWPos3right = IntVec2(640 - 8 - 48, statWPos3left.y)
    }
}
